pkg/kubelet/kubelet.go

```diff
@@ -779,20 +779,29 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 	klet.statusManager = status.NewManager(klet.kubeClient, klet.podManager, klet)
 
 	if utilfeature.DefaultFeatureGate.Enabled(features.RotateKubeletServerCertificate) && kubeDeps.TLSOptions != nil {
-		var ips []net.IP
-		cfgAddress := net.ParseIP(kubeCfg.Address)
-		if cfgAddress == nil || cfgAddress.IsUnspecified() {
+		var (
+			ips   []net.IP
+			names []string
+		)
+
+		// If we successfully discovered addresses from the cloudprovider, use
+		// those. Otherwise, use the configured addressed if provided. Otherwise,
+		// make a best guess.
+		if len(ips) != 0 || len(names) != 0 {
+			ips = cloudIPs
+			names = cloudNames
+		} else if cfgAddress := net.ParseIP(kubeCfg.Address); cfgAddress != nil && !cfgAddress.IsUnspecified() {
```

Comment:
I would have expected the explicitly configured address to override the cloud provider discovered ones. What's the motivation for having discovered addresses override the explicit ones?

Reply:
Ya, explicitly configured addresses should probably come first.